staking/programs/staking/src/state/positions.rs

```diff
@@ -294,4 +312,103 @@ pub mod tests {
         let buffer = [0u8; POSITION_BUFFER_SIZE];
         assert!(Option::<Position>::try_read(&buffer).unwrap().is_none());
     }
+
+    // A vector of DataOperation will be tested on both our struct and on a HashSet
+    #[derive(Clone, Debug)]
+    enum DataOperation {
+        Add(Position),
+        Modify(Position),
+        Delete,
+    }
+
+    // Boiler plate to generate random instances
+    impl Arbitrary for Position {
+        fn arbitrary(g: &mut Gen) -> Self {
+            return Position {
+                activation_epoch:       u64::arbitrary(g),
+                unlocking_start:        Option::<u64>::arbitrary(g),
+                target_with_parameters: TargetWithParameters::VOTING,
+                amount:                 u64::arbitrary(g),
+            };
+        }
+    }
+    impl Arbitrary for TargetWithParameters {
+        fn arbitrary(g: &mut Gen) -> Self {
+            if bool::arbitrary(g) {
+                TargetWithParameters::STAKING {
+                    product:   Pubkey::new_unique(),
+                    publisher: Publisher::SOME {
+                        address: Pubkey::new_unique(),
+                    },
+                }
+            } else {
+                TargetWithParameters::VOTING
+            }
+        }
+    }
+    impl Arbitrary for DataOperation {
+        fn arbitrary(g: &mut Gen) -> Self {
+            let sample = u8::arbitrary(g);
+            match sample % 3 {
+                0 => {
+                    return DataOperation::Add(Position::arbitrary(g));
+                }
+                1 => {
+                    return DataOperation::Modify(Position::arbitrary(g));
+                }
+                2 => {
+                    return DataOperation::Delete;
+                }
+                _ => panic!(),
+            }
+        }
+    }
+
+    impl PositionData {
+        fn to_set(self, next_index: u8) -> HashSet<Position> {
+            let mut res: HashSet<Position> = HashSet::new();
+            for i in 0..next_index {
+                if let Some(position) = self.read_position(i as usize).unwrap() {
+                    res.insert(position);
+                }
```

Comment:
How about adding
else { panic!() }

And then after the loop checking that the rest are None from next_index to the end